\chapter{Conclusions} \label{conclusion-chapter}


\section{Future Improvements to Micromeda that Require Changes in Multiple Components}

As discussed in Section \ref{introduction_summary}, the summary of each chapter of the thesis discusses potential improvements to the component the chapter discusses. However, some large improvements require modifications to multiple components of the Micromeda platform or are outside the scope of previous chapters.

\subsection{Creation of an Automated Pipeline for Rapid and Easy Generation of Micromeda Files}

Currently there is no automated pipeline that users can use to generate Micromeda files. As detailed in Section \ref{micromeda-overview}, to generate Micromeda files users are required to install and run three command-line bioinformatics tools on their organisms' raw genome sequences. For the less technology inclined an automated bioinformatics pipeline should be developed that would first install all the these bioinformatics tools and later run them in series on user supplied genomes. Components of the pipeline could be installed via Conda (\href{conda.io}{conda.io}) and run in Conda environments to prevent these components from interfering with previously installed software. Pipeline automation tools such as Snakemake \cite{koster2012snakemake} or Nextflow \cite{di2017nextflow} could be used to develop such a pipeline. These tools ensure that steps in the pipeline are executed in the correct order and allow for these steps to be scaled out in parallel if multiple organism's genomes are to be processed. Components that are hard to install via Conda, such as InterProScan, could be installed using software containers (see Section \ref{containerization}). Indeed, InterProScan has already been containerized by the author to support the thesis work (see href{github.com/Micromeda/InterProScan\-Docker}{github.com/Micromeda/InterProScan\-Docker}).

An extension of this basic pipeline could also be used speed up the process of generating Micromeda files. As discussed in Section \ref{why-micromeda-files}, InterProScan is very slow as it has to scan an organism's predicted proteins with many domain models. One way of speeding up this process would to be to reduce the set of proteins that are required to be scanned. In Section \ref{Genome-Properties-Files}, it is mentioned that each property in the Genome Properties database posses has an associated FASTA file containing representative protein sequences that carry out the steps of a property. A fast sequence similarity search tool such as MMSeqs2 \cite{steinegger2017mmseqs2} could be used to compare an organism's proteins to the representatives in these FASTA files. This comparison could be used to filter out proteins of an organism that are unlikely to be homologous to those representative proteins from database (i.e., true negatives), which reduces the overall set proteins that would need to be scanned by InterProScan. The aforementioned Micromeda file generation pipeline could automate these filtering steps.

\subsection{Add the Option to Assign Properties According Percentage Completeness Rather Than Categorically.}

Currently the heat map generated by Micromeda is is coloured according to property assignments that follow a discrete scale. Each property is assigned YES, PARTIAL or NO support and is coloured to match. It should be noted that the algorithm that Pygenprop uses to generate these assignments only uses the assignments of required steps for a property in its calculations (see Section \ref{AssignmentCachingAlgorithm}). If even a single required step is not supported, then the parent property is assigned PARTIAL support. If Micromeda is applied to incomplete genomes, such as those generated from Metagenomes, then it is quite possible that many genes encoding for proteins that support some property steps may be missing. These missing genes would result in many properties being assigned PARTIAL support. 

To address these potentially high levels of PARTIAL property assignments created by running Micromeda on incomplete genomes, it may be useful for users to be able to quantify the level of support for each assigned property along a continuous scale. For example, properties that are more of their steps supported could be given higher values than those that are less complete. Such a continuous scale could also be influenced non-required steps. For example, if two organism's possess the same number of required steps for a property but one has a large number of non-required steps supported as well, then it is likely that the latter has more support for the property existing and it could be assigned a hight value. In the context of the heat map, these continuous assignments of support could be assigned different shades of the same colour.

Algorithms for calculating levels of PARTIAL support for biochemical pathways has been developed before. For example, Metabolic and physiological potential evaluator (MAPLE) \cite{takami2016automated}, a KEGG-based pathway annotation server, can calculate Module Completion Ratios (MCR) (i.e., continuous levels of support based on the presence of pathway steps) for each module in the KEGG database and based off of an organism's uploaded genome sequence. KEGG modules are roughly equivalent to higher level genome properties. MCR's are calculations of the level of completeness of individual pathways that take into account that some pathway steps may be catalysed by multiple different enzymes. They are calculated using custom "boolean algebra-like equations" generated for each KEGG module \cite{takami2012evaluation}. These equations take in KEGG Orthology (KO) annotations (see Section \ref{conclusion-chapter}, Fig. \ref{fig:functree2}, and \cite{mao2005automated}) of an organism's proteins as input. KEGG Orthology annotations are equivalent to the step evidences used to support genome property steps. 

A similar "completeness ratio" algorithm to that used by MAPLE could also be implemented by Pygenprop. Much of the logic encoded in MAPLE's "boolean algebra-like equations" is already built into the Genome Properties database itself. For example, the Genome Properties already records how multiple InterPro domains, which representing multiple enzyme families, can be used to support a single step. The steps required for individual properties are also codified. Calculation of "completeness ratios" for genome properties may be complicated by the fact that some genome properties rely on the presence of others (see Section \ref{genome-properties-overview}), unlike KEGG modules, which are only supported by KO annotations.

Such a "completeness ratio" based assignment system could be implemented by giving Pygenprop a second set of functions for calculating genome property assignments. The library could be modified to optionally calculate either discrete or continuous assignments for individual properties. These continuous assignments could be stored in Micromeda files or generated dynamically. In the context of Micromeda, its interface could be modified to have a user interface switch that would allow users to switch their current heat map between displaying property assignments as either YES, NO, or PARTIAL or as a percentage of step completeness. Micromeda-Server would have to be modified to provide an endpoint (Section \ref{endpoints}) for these serving completeness-based property assignments to Micromeda-Client.

\subsection{Expansion of the Genome Properties Database to Included New Properties Not In InterPro}

Although the Genome Properties database has been shown to have similar overall genome coverage to other pathway databases (see Section \ref{reason-for-genome-properties-selection}), it still missing some pathways that are covered by other databases and pathways that are newly emerging in literature. These pathways could be rapidly added to the Genome Properties database by adding a series of custom hidden Markov models (HMMs) \cite{eddy2011accelerated} to InterProScan's data files. These HMMs would be used to detect the enzymes that carry out these new pathway's steps. For example, the FunGene \cite{fish2013fungene} database contains a series of custom HMMER \cite{eddy2011accelerated} HMMs that are used for identifying enzymes involved in key biogeochemical processes. Integrating FunGene as a new InterProScan member database would only require adding FunGene's HMMs to InterProScan's data files, writing a Python script for filtering out false positive HMM matches to these HMMs, and modifying some of InterProScan's configuration files. Afterwards, InterProScan would automatically use these new HMMs when annotating novel proteins and would add the filtered results from using these HMMs to its output TSV file (see Section \ref{overview-interproscan}). This modified version of InterProScan could then be maintained separately or merged into the main InterProScan Github repository. Once FunGene is added to InterProScan, new Genome Properties could be created that used FunGene HMM names as step evidences instead of InterPro signature accessions. This could be implemented by added new property \textbf{DESC} files (Section \ref{Genome-Properties-Files}) to the database. This custom version of the Genome Properties database could also be maintained as a fork or merged into the latest version of the Genome Properties database on Github. A similar design pattern could be used to integrate any HMM-based pathway database. The main advantage of this approach is that it allows the new database's HMMs to be ran by InterProScan, which provides high scalability across many CPU cores or even on HPC clusters.
