\chapter{Development of a Web-Based Visualization Tool for the Comparison of Organism Genome Properties} \label{micromeda-client}

As discussed in Chapter XXX, Micromeda's user interface is provided by a client web application. This client's role is to provide users with a streamlined interface for visualizing patterns of property and step assignment found across organisms. This assignment data is provided to the client in the form of a user uploaded Micromeda file (Section \ref{MicromedaFiles}). After upload, these files are sent to Micromeda-Server (Chapter \ref{micromeda-server}) where they are parsed and used to provide data to the client. The client can request this data through series of web application programming interface (API) endpoints (Section \ref{endpoints}) provided by Micromeda-Server. In this chapter we will discuss the client web application, called Micromeda-Client, in detail.

\section{Visualisation Design} \label{visualization-design}

One of the core uses of Genome Properties assignment data is to mine it for biologically relevant patterns. These patterns are found in the presence and absence of biochemical pathways or structural features across organisms. One of the best ways to detect such patterns is through the use of data visualisation. By visualising such data we can make many types of comparisons between organisms, pathways or steps. Some examples comparisons and their research relevance are listed below and displayed in Fig. \ref{fig:client-analysis-types}. 

\begin{itemize}
\item Looking at the assignments of a single property across organisms can be used to used for the selection of subsets of organisms which may posses a specific phenotype (Fig. \ref{fig:client-analysis-types}a). In the field of genetic engineering, such comparisons useful host and gene donor selection.
\item Comparing the assignments of multiple properties could be used for identifying patterns of property conservation across organisms in a dataset (Fig. \ref{fig:client-analysis-types}c). In microbial ecology, such comparisons could be used to identify microorganisms which fit specific ecological roles.
\item Looking at the steps assignments of a single property in a single organism could be used to evaluate the correctness of the  assignment (Fig. \ref{fig:client-analysis-types}b). For example, a property might be assigned NO or PARTIAL because it is only missing a few required step assignments but possess many others which are not required (Fig. \ref{fig:client-analysis-types}b). Being able to look at all step assignments for the property may be useful for determining why a property has been given an assignment.
\item Comparing the step assignments of single property across organisms can be used to see what pathway steps are retained across organisms that differ photogenically (Fig. \ref{fig:client-analysis-types}d). For example, if a property step is not retained in a large assortment of genomes, it may not be required for a pathway or is carried out by proteins which are non-canonical (Fig. \ref{fig:client-analysis-types}d). Such non-canonical proteins may not posses domains used by Genome Properties but may still carry out the pathway step.
\end{itemize}

\begin{figure}[!ht]
  \centering
	\includegraphics[width=\textwidth]{media/analysis_types.pdf}
	 \caption{Visualizations of Genome Properties data can be used to compare property and step assignments across organisms.}
	 \label{fig:client-analysis-types}
\end{figure}

If Micromeda-Client is to support the above comparison's, the data visualization method used Micromeda-Client should allow users to perform the following tasks.

\begin{itemize}
\item Track assignments across organisms.
\item Assess the magnitude of given assignments.
\item Aggregate step and property assignments into summaries.
\item Explore how these aggregate assignments are derived.
\item Quickly find assignments of interest.
\end{itemize}

When a visualization approach was select for use by Micromeda-Client, the compatibility of the visualization with these tasks was kept in mind.

\subsection{An Overview of Genome Properties Data}

Different types of data are more suitable to specific visualisation techniques and it is important to first discuss the nature of the data to be visualized before discussing why the author selected a specific visualisation type for Micromeda-Client. At its core, the data presented by Micromeda-Client consists of assignments for genome properties and their steps. Such assignments are ordinal data \cite{richardson2018genome,agresti2010analysis}, as their states of YES, PARTIAL and NO, though categorical, are ordered. Its should be noted that properties are also connected to each other in hierarchical manner \cite{richardson2018genome}. This structure can be considered hierarchical data \cite{richardson2018genome,samet1990applications}. The ordinal assignment data is also influenced by property hierarchy as the assignments of parent genome properties can be used to summarize the assignments of child genome properties or property steps \cite{richardson2018genome}. Each piece of assignment data found in a Micromeda file belongs to a specific property or step and organism. Thus, the genome property assignment data can be also considered multidimensional \cite{pedersen1999multidimensional}.

\subsection{How Assignment Data Is Visualized By Micromeda-Server}

When designing a data visualization, sometimes specific visualization techniques immediately stick out. This was the case while designing Micromeda-Client. A very apt visualization for multidimensional datasets, such as Genome Properties assignments, is a heat map \cite{wilkinson2009history,tufte2001visual}(Fig. \ref{fig:client-analysis-types}) and this is the visualization technique chosen for the client (Fig. \ref{fig:client-analysis-types}). In such a visualization, cell position is used to indicate which assignment belongs to what property or step and organism (Fig. \ref{fig:client-analysis-types}). Because most Micromeda files have less organisms than genome properties, we chose to position assignments for the same property in the same heat map row. Each assignment is from a different organisms in the dataset. This configuration leads to a heat map that is much taller than it is wide, which forces users scroll vertically. Scrolling vertically is much more convenient than horizontally as it allows the user to scroll using their mouses scroll wheel. Columns are used to group assignments from the same organism across properties and steps. The magnitude of each assignment is encoded using cell color (Fig. \ref{fig:client-analysis-types}). Since assignments are ordinal data it makes sense to encode assignments using colour saturation, rather than hue \cite{munzner2015visualization}. For Micromeda-Server the author chose to use a purple cell color scheme to insure diagram is interpretable by those with colour blindness. A web service called color brewer (\href{colorbrewer2.org}{colorbrewer2.org}) was used to select these colors and a tool called Sim Daltonism (href{github.com/michelf/sim-daltonism}{github.com/michelf/sim-daltonism}) was used to ensure the color blind compatibility of the entire diagram. A heat map was chosen over competing visualization techniques such as bubble charts \cite{tufte2001visual}, circular maps \cite{ward2002taxonomy,stothard2004circular} or tree maps \cite{shneiderman1998tree} due to the number of variables that needed to be plotted and the superior space utilization of heat maps. Heat maps have superior space utilisation as compared to circular plots because they mimic the square dimensions of the computer monitors they are displayed on (Fig. \ref{fig:circle-square}).

\begin{figure}[!ht]
  \centering
	\includegraphics[width=0.7\textwidth]{media/square_vs_circle.pdf}
	 \caption{A square shape (red), as posses by a heat map, provides better space utilization on a conventional display (grey). The circle (blue) and square have the same area. A circle requires greater X and Y-axis dimensions to display the same amount of data and produces dead space at the displays corners where data cannot be presented.}
	 \label{fig:circle-square}
\end{figure}

Any visualization technique used by Micromeda-Client is also challenged by the shear size of the data to be presented. For example, if all assignments for only a few organisms were displayed in the heat map described above, then its shear size would prevent users to quickly finding assignments or tracking assignments across organisms. As of version 2.0 of the Genome Properties database, there are 1296 properties and 6525 steps. If a single heat map was generated for all properties, with each cell being 5 mm tall, such a heat map would be approximately 6.5 meters tall (i.e., fifteen vertical pages on a 24" monitor). If the same heat map was made for all property steps, the resulting heat map would be over 32.6 meters tall (i.e., seventy-five vertical pages on a 24" monitor). If both of these heat maps were combined, it would be even larger. To address this length issue, Micromeda-Client's visualization interface uses interactive aggregation and de-aggregation of assignment rows to reduce the overall length of its assignment heat map (Fig. \ref{fig:visualization-philosophy}). This reduced length facilitates rapid visual exploration of property and step assignments.

Micromeda-Client's user interface allows users to manipulate the contents of its assignment heat map to show and hide properties and steps according to their position in the Genome Properties Directed Acyclic Graph (DAG) (Fig. \ref{fig:visualization-philosophy}). Because individual properties are arranged hierarchically, the assignments of properties closer to the root of the Genome Properties DAG can be used to summarize the assignments of properties closer to its leafs (Fig. \ref{fig:visualization-philosophy}). In the context of a heat map, this means that a row of assignments for a parent property can be used to summarize the rows of assignments of its child properties. In addition, an assignment heat map row for a leaf genome property can be used to summarize the assignment rows of its steps (Fig. \ref{fig:visualization-philosophy} and Fig. \ref{fig:client-analysis-types}). Micromeda-Client provides mechanisms to expand and collapse heat map rows to display either parent summary assignments or more detailed child assignments (Fig. \ref{fig:visualization-philosophy}).

\begin{figure}[!ht]
  \centering
	\includegraphics[width=\textwidth]{media/visualization_design_philosphy.pdf}
	 \caption{Micromeda-Client's user interface uses a clickable icicle diagram to change the state of an adjacent heat map of property and step assignments. Clicking nodes in the diagram changes the shape of the heat map by adding and removing rows from the diagram.}
	 \label{fig:visualization-philosophy}
\end{figure}

One key design decision for Micromeda-Client was how to let users control the above aggregation and de-aggregation of heat map rows. To support this usage, the author added a new visualisation component in addition to the assignment heat map. Specifically, Micromeda-Client uses a horizontal icicle diagram \footnote{Traditional icicle diagrams have leaf nodes facing downwards. The icicle diagram used by Micromeda-Client is rotated 90 degrees and has it leaf nodes facing to the right.} placed left of the heat map. This icicle diagram is used to control the heat map's content (Fig. \ref{fig:visualization-philosophy}). Icicle diagrams allow for the display of hierarchical data such as the relationship between properties and between properties and steps. Nodes which represent properties are labelled gray and those representing steps supported by protein domains are labelled green. In Micromeda-Client, each of the genome properties and steps in the Genome Properties database is assigned a node in the icicle diagram (Fig. \ref{fig:visualization-philosophy}). The leafs of this icicle diagram are aligned to matching rows in the adjacent heat map (Fig. \ref{fig:visualization-philosophy}). Icicle diagrams were chosen over other ways of visualizing hierarchical data, such as trees, due to their spacial compactness (Fig. \ref{fig:visualization-philosophy}). With Micromeda-Client, the nodes in the icicle diagram are given a state of either on or off. When a node in an off state is clicked, a new column is added to the icicle diagram. This column's contents includes nodes representing the clicked node's children. Simultaneously as the new child columns is added, matching assignment rows belonging to properties or steps that the child nodes represent are added and replace the clicked node's assignment row in the heat map. The clicked node's shape is expanded vertically to align with the top and bottom of its first and last child node cell, respectively. If the clicked node is clicked once again then the process above is reversed, child nodes and their assignment rows are removed from the visualization (Fig. \ref{fig:visualization-philosophy}), the clicked cell returns to its original shape, and its matching summary assignment row is placed back into the heat map (Fig. \ref{fig:visualization-philosophy}). Columns in the icicle diagram are deleted, upon child cell removal, only if all sibling or cousin nodes to the clicked node have no children displayed.

The visualization strategy chosen for Micromeda-client supports the required tasks presented at the top of this section. The heat map allows users to track assignments across organisms and assess their magnitude. The interactive aggregation control provided by the icicle diagram allows for the aggregation of step and property assignments into summaries. These aggregate assignments can be de-aggregated to show child assignments, allowing users to explore how the parent assignment was derived. As the icicle diagram follows the structured of the Genome Properties DAG, specific assignments can be quickly found by following the DAG's structure from parent to child. Searching for properties is further enhanced by Micromeda-Client's ability to search for properties by name. This search functionality is discussed in the next section.

\section{Additional Features of Micromeda-Client's Interface} \label{client-additional-features}

In addition to the visualization capability presented above, the client's interface also possesses several other features which help users explore their data. In the top right corner of the user interface is a text-based search box (Fig. \ref{fig:micromeda-interface}). This allows user's to search for properties by name via entering a text string. As the user enters this string, matching property names are displayed in a drop down menu (Fig. \ref{fig:micromeda-interface}). If one of these property names is clicked, then the Micromeda-Client will automatically scrolls heat map to the row where assignments for the property are located. If the property is not shown in the current version of the heat map, then it will be added by expanding a path to the property. This paths is built via recursively de-aggregating parent properties along a path from the root of the Genome Properties DAG to the property which was clicked in the menu.

The scrolling behaviour above is also activated when a user clicks a node in the icicle diagram. When clicked, the heat map will scroll to align with the top of the clicked node. As the diagram scrolls the X-axis labels remain in a fixed position to provide users with context to which organisms assignments belong. 

\begin{figure}[!ht]
  \centering
	\includegraphics[width=0.9\textwidth]{media/micromeda-interface.pdf}
	 \caption{Micromeda's user interface provides functionality for searching for properties (4), getting additional information about properties and steps (1), and has the ability to download protein sequences which support step assignments (2). A legend provides context to the heat map's colours (3). A reset button allows users to reset the heat map to its initial load state(4).}
	 \label{fig:micromeda-interface}
\end{figure}

While users explore the genome property assignments of organism's in their datasets, it may be useful for them to be able to get more context about the properties whose assignments are displayed. To facilitate this, the icicle diagram has question mark glyphs \footnote{A glyph is a elemental component of a data visualization. A data visualization is composed from a series glyphs. Often glyphs are directly mapped to data points \cite{chen}. For example, the dots of a scatter plot are individual glyphs \cite{chen}.} at the bottom of each node. When these glyphs are hovered over, a pop up box appears displaying information about the property that the node represents (Fig. \ref{fig:micromeda-interface})). The pop up box includes the name of the property, a description of it, a link to the property in the EBI Genome Properties website, and a list of links to equivalent records in other pathways databases such as KEGG \cite{kanehisa2000kegg} and MetaCyc \cite{karp2002metacyc}. When the user's cursor leaves the glyph or the pop box, the box is once again hidden.

Property steps have a different glyph at the bottom of their nodes that look like a download symbol (Fig. \ref{fig:micromeda-interface})). It may be useful for users to be able to download protein sequences that support the existence of property steps across organisms in an uploaded dataset and the the pop up box of this glyph supports that. When the download glyph is hovered over, it opens up pop up box containing two download links (Fig. \ref{fig:micromeda-interface})). The first link, when clicked, downloads a FASTA file containing the protein sequences that are most likely to carry out the step in each organism in a dataset. The second link points towards a second FASTA file containing any protein that could carry out the step in a dataset. When the cursor is removed from this download pop up box, it will be hidden

Micromeda-Client's interface also includes a reset button. This button resets the heat map and icicle diagram back to their original configuration where only top level properties are shown (i.e., one level below the root in the Genome Properties DAG). Its is useful for when users want to reset the diagram to search for other properties.

\section{Delivery Methodology}

Micromeda-Client is delivered as a client web browser application. The method was chosen due to its relative ease of deployment. End users only need open the web address of the client for the client to be loaded into their web browser and run. Since the application is web browser-based, it will work on any operating systems with a modern web browser and even mobile devices such as tablet computers and cell phones.

\section{Implementation}

Micromeda-Client's interface consists of two web pages that were structured using Hypertext Markup Language 5 (HTML5) \cite{HTML5}, styled using Cascading Style Sheets 3 (CSS3) \cite{CSS3}, and scripted via JavaScript (ECMAScript 6) \cite{flanagan2006javascript}. One page is used for uploading user generated Micromeda files and another is used for presenting the visualizations of the file's data to users. To use Micromeda-Client, users must first navigate to the upload page and upload a Micromeda file. After upload is complete, their browse will automatically redirect them to the visualisation page. Both pages are styled using the Bootstrap 3.0 CSS framework \cite{spurlock2013bootstrap}. Bootstrap is used set up page elements such as header navigation bars and drop down menus. Bootstrap also makes each of the pages compatible with tablet computers and phones as it will automatically restyle the pages to fit on these device's smaller screens.

\subsection{Core Data Structures} \label{visual-data-structures}

Two core data structures are used by the client visualization page and they are both used during diagram generation. One is a diagram configuration array which contains a series of measurements which are used during diagram drawing. These measurements control things such as the spacing between heat map cells, heat map cell dimensions and the offset of axis labels. A summary of these measurements can be seen in Fig. \ref{fig:diagram-measurements}. The contents of this setting array is stored in a JSON file \cite{bray2014rfc}, called \textbf{diagram\_configuration.json}, which is deployed alongside the HTML files of the client. The second data structure is a copy of the Genome Properties DAG in the form a tree (i.e., nodes with two parents are duplicated) of JavaScript objects. Each of these objects represents a genome property or step and objects are linked together in parent-child relationships. This data structure is analogous to the one used by Pygenprop (Section \ref{GenomePropertiesTree-Class}).  Each of these objects possesses an attribute, called assignments, containing a list of property assignments for a set of organisms. The objects also have an attribute, called \textbf{enabled}, containing a boolean (i.e., true or false). The contents of the visualization presented by the client  maps directly to the state of property object tree (Fig. \ref{fig:tree-map-to-viz}). When the property tree manipulated by the client and the clients visualization is redrawn, it is redrawn in a way that matches the new structure of the tree. The \textbf{enabled} boolean of each property object tree provides determines whether the children of the property should be displayed in the client's assignment heat map and icicle diagram (Fig. \ref{fig:tree-map-to-viz}). Elements of Micromeda's user interface manipulate enabled states of objects in the property tree and this in turn can manipulate the contents of the visualization (Fig. \ref{fig:tree-map-to-viz}). The property tree is placed within a parent JavaScript object. This object also possesses methods for dictionary style look ups of properties from the tree based on their property identifier and selection root and leaf properties, which is analogous to Pygenprop's GenomePropertiesTree class (Section \ref{GenomePropertiesTree-Class}).

\begin{figure}[!ht]
  \centering
	\includegraphics[width=\textwidth]{media/diagram_measurements.pdf}
	 \caption{A series of spacing, width and length values are used by Micromeda-Server to build its diagrams. These values are stored in an external file. The contents of this file can be modified to change the the layout of Micromeda-Client's visualizations.}
	 \label{fig:diagram-measurements}
\end{figure}

\begin{figure}[!ht]
  \centering
	\includegraphics[width=0.7\textwidth]{media/how_tree_state_maps.pdf}
	 \caption{The state of enabled attributes in the client's Genome Property tree directly maps to contents of its visualization.}
	 \label{fig:tree-map-to-viz}
\end{figure}

\subsection{Loading the Visualization}

As mentioned previously, much of the functionality of Micromeda-Client is supported by requests for data from Micromeda-Server. All these requests are done through a technique known as Asynchronous JavaScript and XML (AJAX) \cite{garrett2005ajax,li2012jquery} (see \href{en.wikipedia.org/wiki/Ajax\_(programming)}{en.wikipedia.org/wiki/ Ajax\_(programming)}). AJAX allows requests to be made to the server, using JavaScript \cite{flanagan2006javascript}, without the need for a web page reload. This technique allows Micromeda-Client to remain asynchronous and not kept in sync with the server to where it was downloaded. AJAX requests are made via Hypertext Transfer Protocol (HTTP) \cite{fielding1999hypertext}, through a series of Uniform Resource Locator (URL) addresses \cite{berners1994rfc} which map to Micromeda-Server endpoints (Section \ref{endpoints}). All AJAX requests to the server were made using the JQuery JavaScript library \cite{chaffer2013learning,li2012jquery}. The URL address of Micromeda-Server, where AJAX requests are made to, is found in a JSON file called server\_config.json, which is deployed with HTML files of the client. This file is loaded by both the file upload and visualization page upon their initial load.

The upload page contains a file drag and drop zone and when a user drops a Micromeda file on this zone it is sent, via AJAX, to Micromeda-Server using its \textbf{upload} endpoint (Subsection \ref{endpoint-upload}). The drag and drop zone was implemented using DropzoneJS \cite{meno}. After the file upload is complete, a dataset key is returned to Micromeda-Client. This dataset key is stored in the browser's web local storage \cite{Hickson} (\href{en.wikipedia.org/wiki/Web\_storage}{en.wikipedia.org/wiki/Web\_storage}) using a library called localForage \cite{localforage} \footnote{LocalForage (\href{github.com/localForage/localForage}{github.com/localForage/localForage}) is a wrapper library for a variety of web browsers' local storage systems. It provides Micromeda-Client with compatibility with broad range of browsers including those with more outdated local storage systems such as the one possessed by older versions of Safari \cite{lawson_2014}.} Once the dataset key is stored, the browser is redirected to the visualisation page where the visualization of uploaded file's contents is generated.

As the visualisation page loads, the client requests a JSON tree from Micromeda-Server's \textbf{get\_tree} endpoint (Subsection \ref{get-tree}) via AJAX. The dataset key saved in local storage is provided with this request and ensures that data from the recently uploaded Micromeda file is returned. Micromeda-Server will respond to the request to the \textbf{get\_tree} endpoint by returning JSON tree containing assignments for all properties and steps within the uploaded Micromeda file. This JSON file is parsed into a JavaScript tree, which is used as the core property tree data structure mentioned in Section \ref{visual-data-structures} above. The visualization is then built using the dimensions from the diagram configuration array, the contents of the property tree and the \textbf{enabled} attribute of the property objects therein. The visualisation are generated using version 3.0 of the D3.js visualization library \cite{bostock2015d3} and custom JavaScript library.

\subsection{Interactivity After Initial Visualization Load}

One of the key properties of clients visualizations is that it is interactive and this interactivity is provided by manipulating the JavaScript property tree. When a property node in the icicle diagram is clicked, an JavaScript DOM (Document Object Model) onclick event \cite{dom-events} (\href{en.wikipedia.org/wiki/DOM\_events}{en.wikipedia.org/wiki/DOM\_events}) is triggered, which changes that  state of equivalent property object in the core property tree data structure. Specifically, the clicked node's equivalent property object's \textbf{enabled} attribute is inverted upon icicle node click (Fig. \ref{fig:tree-map-to-viz}) and the entire visualisation is subsequently re-rendered. Because the enable attribute has changed on the property object, the heat map visualization is re-rendered in a different configuration (Fig. \ref{fig:tree-map-to-viz}). 

In the case where a leaf node on the icicle diagram is clicked, the matching property object's \textbf{enabled} attribute is changed from false to true. Then when the diagram is subsequently re-rendered, the children of the clicked node are now rendered (Fig. \ref{fig:tree-map-to-viz}). After this re-rendering, the page is scrolled so that the bottom of the X-axis labels are aligned with the top of the clicked node. The opposite occurs when a non-leaf node on the icicle diagram is clicked. The associated property object's \textbf{enabled} variable is changed from true to false and after re-rendering of this node's children are removed. The page is scrolled so the bottom of the X-axis labels are aligned with the top of the clicked node. The true and false state of child property objects is not changed. This provides the visualization with a sort of "memory". When a child node is re-enabled on the diagram the state of its children is retained (Fig. \ref{fig:tree-map-to-viz}).

At page load the property tree is used to create an JavaScript array of pairs of property names and identifiers. After the visualisation generation, an interactive search menu is created in the top right hand corner of the Micromeda-Client interface (Fig. \ref{fig:micromeda-interface})). This menu is built using the Select2 library \cite{select2} (\href{select2.org}{select2.org}) and uses previously mention pairs of property identifiers and names as search data. When a property name or identifier is searched for in the search menu, properties whose name is similar to the search query are displayed in a drop down menu (Fig. \ref{fig:micromeda-interface})).

When one of these menu properties are clicked then another JavaScript event occurs where the parent of the matching property object in the JavaScript property tree is modified. All property objects along the tree in a path from the root to the parent of matching property object have their \textbf{enabled} attribute set to true. Thereafter, when client visualization is subsequently re-rendered, a path to the assignments of the clicked property is displayed along with the property's assignments. The page is scrolled so the bottom of the X-axis labels are aligned with the top of the clicked node.

After each re-rendering of the assignment visualization, a series of AJAX requests sent. These are requests for information about each property that is visible in the icicle diagram (Fig. \ref{fig:micromeda-interface})). These requests are sent to Micromeda-Server's \textbf{Get\_Single\_Genome\_Property \_Info} endpoint (Subsection \ref{get-property-info-endpoint}). The endpoint returns a JSON document containing the information about a property, which is later used in property information pop up boxes (Fig. \ref{fig:micromeda-interface})). The contents of these documents are also cached in web local storage using localForage \cite{localforage}. During subsequent diagram re-renderings requests are not made for properties whose data is already cached. This caching reduces request frequency to the server and overall server load. The request volume is also reduced by only making requests  for properties that are visible.

Both property information and FASTA download pop up boxes are generated upon activation of JavaScript onmouseover events \cite{dom-events} (\href{en.wikipedia.org/wiki/DOM\_events}{en.wikipedia.org/wiki/DOM\_events}) of the property question mark and step download glyphs, respectively. When the a question mark glyph of a property icicle diagram node is hovered over, information about the property is retrieved from the web local storage using localForage. This property information was cached during the last time the diagram was rendered, and a template of this information is used if no information for a specific property is found. The property information is used to generate the contents of property information pop up box. The FASTA download pop box, when generated contains links for downloading FASTA files. These links point to URLs provided by Micromeda-Server's \textbf{get\_fasta} endpoint (Subsection \ref{get-fasta-endpoint}). When the get all proteins link is clicked (Fig. \ref{fig:micromeda-interface})), the endpoint URL's \textbf{all} HTTP GET parameter is set to true during the request and thus FASTA file containing all proteins that support the step is downloaded (see Subsection \ref{get-fasta-endpoint}). The \textbf{all} HTTP GET parameter is not added during requests initiated by clicking the get top proteins link. The lack of this parameter leads to the downloaded  FASTA file containing only the proteins that are most likely to support a step. During both of these requests, both for all and only the top proteins, the dataset key stored in web local storage is provided. This key tells the Micromeda-Server to generate FASTA files from protein sequences stored in the recently uploaded Micromeda file. Both pop boxes are hidden upon removal of the users cursor. This hiding functionality is triggered by a onmouseout event \cite{dom-events} (\href{en.wikipedia.org/wiki/DOM\_events}{en.wikipedia.org/wiki/DOM\_events}).

\section{Interface Performance}

It was found that, other then those that wait for data from Micromeda-Server, all interactive components of Micormeda-Client had almost no visual lag attributed to their construction. For example the visualisation was capable of being re-rendered almost instantaneously even when several hundred rows of assignments were to be displayed.

\section{End-User Testing}

Three potential user of Micromeda-Client were provide with visualization of data relevant to their research project. These users were interviewed afterwards and their suggestions for improving the clients user interface recorded. Some of their suggestions are discussed in Section \ref{client-improvements}. 

\section{Deployment}

Micromeda's user interface can be deployed in two ways. It can be served from a web server, such as NGINX \cite{reese2008nginx}, running on the same server computer system as Micromeda-Server (Fig. \ref{fig:client-deployment}). Such a web server is already of component of the medium scale deployment of Micromeda-Server described in Subsection \ref{single-server-micromeda-deployment}. It can also be served from a Content Delivery Network (CDN) \cite{farber2003internet} (\href{en.wikipedia.org/wiki/Content\_delivery\_network}{en.wikipedia.org/wiki/Content\_delivery\_network}) such as Amazon Cloudfront \cite{varia2014overview} (\href{aws.amazon.com/cloudfront}{aws.amazon.com/cloudfront}) or Cloudflare Anycast \cite{calder2015analyzing} (\href{cloudflare.com/cdn}{cloudflare.com/cdn}). In this deployment configuration, the end user downloads the client code from nearest content delivery server in the the CDN, rather than the same sever as to where Micromeda-Server is hosted (Fig. \ref{fig:client-deployment}). After loading, the client would then still send API requests to Micromeda-Server, which would be hosted outside the CDN (Fig. \ref{fig:client-deployment}). During end user testing the author deployed a version of Micromeda-Client via Amazon Cloudfront CDN. It was shown to improve loading speed of the HTML, CSS and JavaScript files to the client by approximately 20\%. For development purposes the files for Micromeda-Client could also be downloaded and opened by a developer's web browser for testing purposes.

\begin{figure}[!ht]
  \centering
	\includegraphics[width=0.7\textwidth]{media/micromeda-client-deployment.pdf}
	 \caption{The client files for Micromeda-Client, in production, can either be deployed on the same server computer system as Micromeda-Server (A) or via a content delivery network (B).}
	 \label{fig:client-deployment}
\end{figure}

\section{Future Improvements} \label{client-improvements}

Several improvements could be made to the client to improve its overall usefulness. These include functions that allow make the client easier to user and provide user with more information. Several of these poptential improvements were derived from user feedback gathered during user testing.

\subsection{Providing Information About Property Steps and Improved Search Capability}

In the current version of Micromeda-Client, there are pop boxes that provide information about additional individual genome properties such as descriptions of the property and links to equivalent records in other pathway database. However, there are no equivalent pop up boxes for providing detailed information about property steps. The Genome Properties database also includes additional information about steps that could be included in step information pop up boxes. For example, information about which domains support a step could be used to generate information pop up box links to records on the InterPro website that detail these domains. Additionally, some steps are assigned Gene Ontology (GO) terms and links to information about these terms on the GO website could also be provided in such pop up boxes. Such information would provide users with additional context about the steps assignments for which they are looking at. Such pop up boxes could be activated by hovering over a question mark glyph placed slightly above the download glyph of each step node in the clients assignment visualization (Fig. \ref{fig:micromeda-interface}).

Currently, the search menu in the top right corner of the client interface allows for users to search for genome properties by name or identifier (Section \ref{client-additional-features} and Fig. \ref{fig:micromeda-interface}). The ability to search for properties could be expanded by allowing users to search for properties by the identifier of equivalent records from KEGG or MetaCyc. This would allow users, who are familiar with the identifiers of specific KEGG or MetaCyc pathways, to rapidly find the equivalent pathway genome property in the Micromeda-Client heat map. It may also be useful to be able to search for property steps, rather than just properties, by name and have the visualization open up a path and scroll to the assignment row for a searched step. Steps could also be searched for by supporting InterPro domain signature identifier or GO term number. Both the addition of step information pop up boxes and improved ability to search for properties and steps would require additional endpoints to be added to Micromeda-Server.

\subsection{Improving the usability of Visual Interface}

During end user testing some users suggested that it was difficult to determine if nodes in the icicle diagram were clickable and which had been previously aggregated and deaggregated. One way of visually declaring whether a node is aggregated and deaggregated is to a '+' sign glyph to the top of each aggregated icicle diagram cell and then replace swap this glyph with a '-' glyph upon deaggregation. If this cell is once again aggregated, then the '-' glyph would once again be changed back to a '+' to indicate that the glyph is once again available to be deaggregated.

A variety of JavaScript libraries have been built to help introduce users to features of client web interfaces. Such libraries, such as intro.js \cite{mehrabani} (\href{introjs.com}{introjs.com}), generate pop ups around interface components that walk the user through and describe each component of this interface. Such a library could be added by Micromeda-Client to assist user in knowing how different interface components work, such as clicking icicle diagram cells.

\subsection{Automating the Scaling of the Visualization for Different Screen Sizes}

TODO

\subsection{Adding Glyphs to Indicate Which Steps Are Required}

One of the core tasks of Micromed-Client, outlined at the top of Section \ref{visualization-design}, was to help users explore how assignments for properties were derived. As described in Section \ref{AssignmentCachingAlgorithm}, Pygenprop uses the required attribute of steps to perform its assignment calculations. However, the client visualization currently provides no indications of what properties or step are required versus which are not. In end-user testing, potential users asked for properties and steps that are required for a parent's assignment be labelled. To solve this problem in Micromeda-Client glyphs could be added to properties and steps that are required versus those that are optional. The \textbf{get\_tree} endpoint of Micromeda-Server (Subsection \ref{get-tree}) could be modified to label properties and steps in its returned JSON property tree with an required attribute. This attribute could be set to true or false depending on whether the property is required or not. 

\subsection{Modification of Micromeda-Client to Collect Assignment Data From a Second Endpoint}

Currently, Micromeda-Client gathers its assignment data from the \textbf{get\_tree} endpoint of Micromeda-Server (Subsection \ref{get-tree}). There is problems with this approach and it would more appropriate to have a separate server endpoint for returning assignments for properties and steps, as discussed in Subsection \ref{assignment-endpoints}. Having the client make a separate API call to gather this assignment data could be useful as it would allow the client to very easily request for the data to be rearranged before it is sent. It would also allow for step assignments that are supported by protein domains to be replaced by match E-value scores.

\subsection{Clustering Heat Map Columns by Assignment Similarity}

Columns in the heat map display contain assignments from different organisms. Currently, these columns are ordered alphanumerically by organism name. It may be useful for users to be able to cluster the columns by assignment contents rather than by name. Clustering by assignment contents would group organisms that have similar assignments and potentially by similar ecological role as organisms with similar ecological roles should have similar property assignments. Columns could be clustered either globally by the assignments of all properties and steps or locally by only those properties and steps that are visible. The simplest solution for clustering these columns would be for Micromeda-Server to perform clustering. As noted in Subsection \ref{assignment-endpoints}, if property and step assignments were pushed to a separate endpoint they would be generated by serializing assignment DataFrames of GenomePropertiesResultsWithMatches to JSON. If clustering columns in the heat map was required, it could be accomplished by clustering these DataFrames column-wise using sci-kit learn. These clustered DataFrames would then be converted to JSON and sent to the client if requested.

\subsection{Visualizing Step Assignments by E-value Score Rather Than by Categorical Assignment}

Leaf genome properties are supported by steps that use InterProScan matches to InterPro consortium domain signatures as evidence. These matches, between a motif found in a protein of an organism and a InterPro domain, has an E-value score. These scores are stored in Micromeda-files. Matches with E-value scores that are below a specific per-InterPro member database threshold are automatically filtered out by InterProScan. Matches that remain are likely to be true positives (i.e., the motif matched is an ortholog domain from the database). However, there is still some E-value score variation among the remaining matches. Motifs with matches that have lower E-value scores are closer in sequence to domains in the database and are more likely to be orthologous. These lowest E-value matches could be said to be the "strongest" matches.

During end user testing, some potential users were interested in having a way to compare the relative strength of step assignments between organisms. For example, if two organisms are assigned YES, which is more likely to possess a step. Step assignments are currently assigned a binary YES or NO and thus the relative strength of these assignments cannot compared. One way to compare the strength of their supporting domain matches. For example, the E-value score of the closest match supporting each step could be displayed in the heat map in place of a binary YES or NO value. These E-value scores could be coloured by shades of green along a continuous scale. Pop up boxes could be generated by hovering over each cell in the heat map cell and would display match info, such as the E-value score, protein name and matched domain identifier. No E-Value data would be presented for NO assignments. A interface switch could be implemented that would be used to switch property assignments between binary YES and NO value and continuous E-value scores. A Micromeda-Server endpoint would have to be built to provide these E-value score data, as discussed in Subsection \ref{e-value-endpoint}.

\subsection{Improving the search and filtering capability of the Client}

During end user testing, it was also identified that some users wanted a way to quickly identify assignments that differ between organisms. This could be done by removing assignment rows from heat map that possess the same assignment across all organisms in the dataset, leaving only those that differ. A user interface switch would have to be implemented that would allow users to switch between a complete assignment view of all assignment rows and only those that differ. The assignment rows would have to be dropped based on those which are displayed in the currently displayed in the heat map. Each node in the JavaScript property could be given an property called \textbf{differing} that returns true if the property's associated assignments are differing between organisms and false otherwise. Children of nodes on the JavaScript property tree whose state is set to \textbf{enabled} would only be rendered if their own \textbf{differing} property is set to true.

Users also asked to be able to re-root the icicle diagram based on the assignments and child assignments of a specific property. For example, having the visualization only show iron metabolism properties and its children. This feature could be implemented by having a glyph each node of the icicle diagram each property that allows the diagram to rooted at the node clicked. In the future, this feature could be expanded to allows users to select a few properties to display. For example, by clicking three of the above glyphs and and interface button.

\section{Summary} 

The web client of Micromeda allows users to visually explore and compare assignments for genome properties and their steps across organisms. It also provides information about these properties and steps and provides links between them and equivalent records in other databases. Due to the use of Micromeda files, it also allows users to download protein sequences that support the displayed assignments. Several web based software tools already exist for visually comparing the presence and absence of biochemical pathways across organisms. However, what sets Micromeda-Client's apart from these tools is its level of data integration in a single view. 

The Genome Properties website has a property viewing program that displays assignment heat maps similar to those of Micromeda-Client (Fig. \ref{fig:property-viewer}). However, information about the properties displayed is not accessible from the heat map and must be viewed in a second web page called a property browser (Fig. \ref{fig:property-browser}). This browsing page is not searchable and users must manually search for each property they want to learn about. Even when a property is found in this browsing page a third page must be opened that contains the property's information. If users need a more detail description of a property, they are forced to swap between between multiple web pages. A web browser window containing the website's assignment heat map could be placed alongside a second window containing the site's properties browser. However, this requires either a very large monitor or two monitors as the heat map page of the website does not scale well to smaller bowser windows. The Genome Properties website's assignment viewer also only displays a subset of leaf properties and does not perform any sort of aggregation of assignments.

\begin{figure}
     \centering
     \begin{subfigure}[b]{0.46\textwidth}
         \centering
         \includegraphics[width=\textwidth]{media/genome_properties_viewer.png}
         \caption{The Genome Properties viewer generates heat maps based on the pre-calculated assignments for a series of reference organisms. Users can also upload InterProScanTSV files for assignment calculation and display.}
         \label{fig:property-viewer}
     \end{subfigure}
     \qquad % some horizontal space
     \begin{subfigure}[b]{0.46\textwidth}
         \centering
         \includegraphics[width=\textwidth]{media/genome_properties_browser.png}
         \caption{The Genome Properties browser provides information about individual properties and their hierarchies.}
         \label{fig:property-browser}
     \end{subfigure}
     \caption{The Genome Properties website provides two separate views for learning more about individual genome properties and viewing property and step assignments. With Micromeda-Client these two views are integrated together into a single joint visualization.}
     \label{fig:genome-properties-interface}
\end{figure}

The issue of having to swap between multiple web pages is an omnipresent problem with many pathway annotation websites. For example, Microscope, a pathway annotation website where users can upload genomes, also possess a heat map conveying KEGG or Metacyc pathway presence and absence. For this website, KEGG Orthology annotations of proteins generate assignments of YES or NO assignments for KEGG modules and pathways. Like Genome Properties these modules and pathways from a hierarchy. Microscope use a similar assignment aggregation and deaggregation technique as Micromeda-Client. However, rather than a child pathway assignment being displayed within the same visualization view, an entirely new heat map is generated for the child properties in a separate page. When navigating from high level pathways to individual KO assignments, users are required to open several heat map pages. In addition to opening each level of the visualization in a separate page, if users need to find more information about a pathway then need to click on links that take them to a page on the KEGG website that describes the property. There is no way to compare the child pathway assignments of multiple high level pathways within the same heat map or page.

\begin{figure}[!ht]
  \centering
	\includegraphics[width=0.9\textwidth]{media/microscope.png}
	 \caption{The MicroScope annotation server can show the presence and absence of pathways across organisms. Figure is from \cite{vallenet2016microscope}.}
	 \label{fig:microscope}
\end{figure}

While interviewing potential potential users of Micromeda-Client, they conveyed their frustration with many other pathway annotation visualization tools and how they force them to navigate through many web pages to gather the information they need for their analysis. They mentioned that it took them much time to swap between pages and, in the case of Microscope, compare assignments across pathway steps. 

In response to this information, Micromeda-Client was designed to integrate as much information as possible into a single view. Interactivity is used to show and hide information as to not overwhelm users. With Micromeda-Client, pop boxes that are convey property information are part of its assignment visualization. Thus users are not required to switch between a series of web pages to learn more about the properties displayed. In addition, with the client, the hierarchy of pathway assignments is displayed within the same view.  The interactive aggregation and deaggregation of Micromeda-Client's heat map rows allows users to rapidly compare child property assignments of sibling parent properties unlike the heat maps presented by Microscope. The icicle diagram and aggregation features of Micromeda's visualization organizes assignment rows and helps users find properties of interest, unlike the Genome Properties website whose heat map only presents assignments for leaf properties.

\begin{figure}[!ht]
  \centering
	\includegraphics[width=0.9\textwidth]{media/functree2.png}
	 \caption{In Functree2, leafs of a radial tree represent individual KEGG Orthology (KO) annotations. In other words, each KO leaf node represents a type of protein that can be found in a cell. Nodes closer to the root represent different levels of classification that can group these KO annotations. Each leaf node is annotated by a stacked bar chart representing the sum of the proteins with the KO annotation that the node represents across all organisms in a dataset. Each color in these radially oriented bar charts is associated with an organism in the dataset. stacked Bar charts also annotate other nodes in the radial tree. The stacked bar charts on higher level nodes represent reciprocal summations of KO counts of their child nodes. Nodes can be clicked to remove child nodes and change the shape and size of the visualization. Figure is from \cite{darzi2019functree2}.}
	 \label{fig:functree2}
\end{figure}

In contrast to the Genome Properties website and Microscope, some alternative annotation visualization tools do present their data hierarchically within the same visualization.  For example, Functree2 allows users to plot KO annotation frequency (e.i., the number of proteins of an organism that that possess a specific KO annotation) across multiple organisms (Fig. \ref{fig:functree2}). Frequencies are visualized using a radial tree with nodes annotated by stacked bar charts. The bar charts of leaf nodes of the tree are represented by counts for a singular KO whereas charts annotating nodes closer to root display summed frequency counts for child nodes (Fig. \ref{fig:functree2}). Unlike Micromeda-Client, aggregated and deaggregated data is presented simultaneously. Nodes in the tree can be clicked on to add and remove child nodes from the visualization. Instead of a default KEGG-based tree, users can also upload their own tree and annotation information.

In the application note for Functree2 its authors note that they chose to use a radial tree over horizontal trees in their visualization to save space. Radial trees are more space efficient than horizontal trees in terms of horizontal space utilized. Also space must provided for the inner bar charts. Rather than defaulting to a horizontal tree or radial tree, micromeda's client replaces both of these trees with an icicle diagram. This icicle diagram provides superior spacial compactness to either tree type. In addition, Micromeda-Client only allows users to show it the aggregated assignments for a node or the deaggregated assignments of its children, not both simultaneously. Both these decisions allow the client's diagram to retain the same square layout as an annotated horizontal tree with superior space utilization to a radial tree. When comparing radial sunburst (as used by tools such as XXX) and rectangular icicle (as used by) space filling data visualization layouts it has been found that icicle diagrams allow users to have higher accuracy and efficiency when finding data. The Micromeda's author asserts that, radial layouts are most useful when they show connections between data on opposite sides of the circle. This design pattern is used by visualization tools such as Circos \cite{krzywinski2009circos}. 