\chapter{A Overview of the Genome Properties Database, InterPro and InterProScan5} \label{genome-properties-chapter} \label{genome-properties}

The architecture and implementation of Micromeda's components, such as Pygenprop and Micromeda-Client, are closely tied to the structure of data held within the Genome Properties database. Thus it is important to provide a short overview of the database's contents and structure before delving into the details of Micromeda's implementation. This chapter will also provide a brief overview of InterProScan, the tool used to search for markers of the existence of specific enzymes within an organism's predicted proteome.

\section{An introduction to Protein Domains and their use as Markers of Enzyme Function}

A protein domain is conserved part of a protein's sequence that carries out a specific function for the protein and is evolutionarily conserved. These domains often generate discrete three-dimensional structures during folding of their parent protein. An enzyme's function is highly correlated with its domain content. For example, all enzymes has an active site (\href{en.wikipedia.org/wiki/Active\_site}{en.wikipedia.org/wiki/Active\_site}), an area on its three dimensional folded shape, that directly interacts with the substrates (see Section \ref{enzymes-and-pathways}) for which an enzyme catalyses. This active site area has a unique sequence of amino acids, a domain, that is often identifiable to a specific enzyme or a family of genetically related enzymes that catalyse a specific reaction type. Often proteins consist of a series of domains. For example, a protein that embeds itself within a cell membrane will have a domain associated with this capability \footnote{Proteins that have a specific set of amino acids on parts of their surface become hydrophobic and tend to embedded themselves within cell membranes that are also hydrophobic. Protein sequence patterns associated with this trait are easily recognized.} as well as its catalytic domain. Similar to how different enzymes can be swapped in and out of biochemicals pathways, domains can also be gained and lost from their host proteins overtime and according to evolutionary forces. For example, enzymes that are evolutionarily related may locate differently in cell due to the presence or absence of a membrane associated domain. It is quite common to have some members of a enzyme family systolic (i.e., free floating within the cell) versus others membrane bound. Thus, it can be concluded that a proteins domain content can be used to identify its potential cellular role.

\section{Domain Annotation}

Motifs are discrete patterns in a protein's sequence, that are often associated with the existence of a protein domain. Domain annotation is the process of predicting the placement and function of domains in protein sequences. During the genomic analysis, it is common practice to perform domain annotation on an organism's predicted proteins.

Domain annotations of an organism's proteins are created by finding similarities between motifs in an organism's proteins and previously seen motifs of domains from other proteins found in a protein database. If an organism and database motif are highly similar in protein sequence, they are said to form a match. The quality of this match can be quantified by metrics such as expected value (E-value) score. A match's E-value score captures how likely it is that the match is real given the chance of finding an equivalent match randomly in one of the organism's other proteins. If we determine that a match is of high quality, the motif in the organism's protein can be assigned the same name and function as the original domain motif the protein database.

One tool for performing domain annotation of an organism's proteins is InterProScan, which predicts both the type and placement of domains in an organism's proteins and also provides supporting match information to justify its predictions. This supporting information includes E-value scores for matches and predicted domain start and stop points on the annotated protein. InterProScan takes a FASTA file \cite{pearson19905} containing an organism's predicted proteins (as would be created by Prodigal) as input and writes domain annotations and match data to tab-separated value (TSV) files.

InterProScan 